README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Sweet spot analysis

## Find and assess treatment effect sweet spots in clinical trial data

Identifying heterogeneous treatment effects (HTEs) in randomized
controlled trials is an important step toward understanding and acting
on trial results. The method in this package exploits any existing
relationship between illness severity and treatment effect, and
identifies the “sweet spot”: the contiguous range of illness severity
where the estimated treatment benefit is maximized. We further compute a
p-value to compare to the null hypothesis of no treatment effect
heterogeneity, and we bias-correct our estimate of the conditional
average treatment effect in the sweet spot. Finally, we provide a
function for visualizing results. Because we identify a single sweet
spot and p-value, we believe our method to be straightforward to
interpret and actionable: results from our method can inform future
clinical trials and help clinicians make personalized treatment
recommendations.

# Example 1: data with a sweet spot

We generate randomized trial data with a treatment effect sweet spot. We
choose n=1000 trial participants with p=10 covariates that - together
with treatment - determine probability of a positive outcome. Treatment
assignment is randomized: every participant has a 50% chance of
receiving treatment.

A participant is in the “sweet spot” if their probability of a positive
outcome (without treatment) is in the middle 20% of probabilities. The
average treatment effect for participants not in the sweet spot is 5%:
that is, treatment additively increases the probability of a positive
outcome by 0.05. For participants in the sweet spot, this is instead
25%.

``` r
set.seed(1234)
n <- 1000; p <- 10;
treated    <- sample(c(0,1), n, replace=TRUE)
covariates <- matrix(rnorm(n * p), nrow=n, ncol=p)
beta       <- rnorm(p)
outcome.prob <- 1/(1+exp(-(covariates %*% beta)))
in.sweet.spot <- !is.na(cut(outcome.prob, c(.4, .6)))
outcome.prob[treated==1 & !in.sweet.spot] <- outcome.prob[treated==1 & !in.sweet.spot] + .05
outcome.prob[treated==1 &  in.sweet.spot] <- outcome.prob[treated==1 & in.sweet.spot]  + .25
outcome.prob     <- pmin(outcome.prob, 1)
positive.outcome <- rbinom(n, 1, prob=outcome.prob)
negative.outcome <- 1-positive.outcome
```

We are now ready to run the sweet spot analysis.

``` r
result <- sweetspot(treated, covariates, negative.outcome, positive.outcome, "binomial")
plot_sweetspot(result, title="Sweet spot on simulated data")
```

![](man/figures/README-example_with_sweetspot-1.png)<!-- -->

We may also look at the predictions from the prevalidated predilection
score model. For example, we can compute the AU-ROC of the predilection
score model on untreated patients:

``` r
pos.scores <- 1/(1+exp(-result$predilection.scores[negative.outcome==1 & treated==0]))
neg.scores <- 1/(1+exp(-result$predilection.scores[negative.outcome==0 & treated==0]))
auc <- mean(sample(pos.scores,10000,replace=T) > sample(neg.scores,10000,replace=T))
auc
#> [1] 0.827
```

# Example 2: data without a sweet spot

We now generate data without a sweet spot. We again choose n=1000 trial
participants with p=10. The treatment effect for all patients is 5%.

``` r
set.seed(1234)
n <- 1000; p <- 10;
treated    <- sample(c(0,1), n, replace=TRUE)
covariates <- matrix(rnorm(n * p), nrow=n, ncol=p)
beta       <- rnorm(p)
outcome.prob <- 1/(1+exp(-(covariates %*% beta)))
outcome.prob[treated==1] <- outcome.prob[treated==1] + .05
outcome.prob     <- pmin(outcome.prob, 1)
positive.outcome <- rbinom(n, 1, prob=outcome.prob)
negative.outcome <- 1-positive.outcome

result <- sweetspot(treated, covariates, negative.outcome, positive.outcome, "binomial")
plot_sweetspot(result, title="Sweet spot on simulated data")
```

![](man/figures/README-example_without_sweetspot-1.png)<!-- -->

# References

For details of this method, please see the following preprint:

Erin Craig, Donald A. Redelmeier, and Robert J. Tibshirani. “Finding and
assessing treatment effect sweet spots in clinical trial data.” arXiv
preprint arXiv:2011.10157 (2020). <https://arxiv.org/abs/2011.10157>

For more about the data, please see:

Dondorp, Arjen M., et al. “Artesunate versus quinine in the treatment of
severe falciparum malaria in African children (AQUAMAT): an open-label,
randomised trial.” The Lancet 376.9753 (2010): 1647-1657.
[https://www.sciencedirect.com/science/article/pii/S0140673610619241](The%20Lancet)

The data in this package was downloaded from Github:
<https://github.com/Stije/SevereMalariaAnalysis>.

It was initially published to accompany the following research:

Watson, Leopold et al. Collider bias and the apparent protective effect
of glucose-6-phosphate dehydrogenase deficiency on cerebral malaria
eLife (2019). [https://elifesciences.org/articles/43154](eLife)

Leopold, Watson et al. Investigating causal pathways in severe
falciparum malaria: a pooled retrospective analysis of clinical studies
In Press, PLoS Medicine (2019).
[https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1002858](PLoS%20Medicine)

Watson, J.A., Holmes, C.C. Graphing and reporting heterogeneous
treatment effects through reference classes. Trials 21, 386 (2020).
[https://doi.org/10.1186/s13063-020-04306-1](Trials)


NEWS.md
# sweetspot 0.1.0
Initial release of the sweet spot analysis package

R/sweetspot.R
#' Perform a sweet spot analysis on clinical trial data. 
#' 
#' @description Identifying heterogeneous treatment effects (HTEs) in randomized controlled trials is an important step toward understanding and acting on trial results. However, HTEs are often small and difficult to identify, and HTE modeling methods which are very general can suffer from low power. This method exploits any existing relationship between illness severity and treatment effect, and identifies the "sweet spot", the contiguous range of illness severity where the estimated treatment benefit is maximized. We further compute a bias-corrected estimate of the conditional average treatment effect (CATE) in the sweet spot, and a p-value. More information here: \url{https://arxiv.org/abs/2011.10157}. 
#' 
#' @param treated A binary vector with one entry for every individual in the trial. The value 1 indicates that the individual was treated, 
#'  and the value 0 indicates that they were a control.
#' @param covariates A numeric matrix containing rows of covariates for every individual in the trial (in the same order as in `treated`).
#' @param negative.outcome The response we use to compute our predilection score. This will determine the model we use.
#' This is a vector with values that may be binary (logistic regression) or continuous (linear regression). 
#' In the binary case, this may be 1 if the patient died, and 0 else.
#' @param positive.outcome The response we use to compute the treatment effect. In the binary case, this may be 1 if the patient lived, and 0 else.
#' @param family A string indicating the response type. Options are "binomial" (for logistic regression) or "gaussian" (for linear regression).
#' @param regularized Boolean indicating whether to use regularization in the predilection score model. If TRUE, we use the optimal model returned by cv.glmnet.
#' @param control.treated.ratio The (positive integer) number of controls for each treated individual.
#' @param predilection.score.nfolds The (positive integer) number of folds we use in pre-validation for the predilection score. The default is 10 folds.
#' @param ntrials.significance The number of bootstraps to run when computing the p-value. Default is 1000.
#' @param ntrials.bias The number of bootstraps to run when debiasing the estimate of the treatment effect. Default is 1000.
#' @param ncores The number of cores to use when finding the sweet spot. If NULL, defaults to parallel::detectCores()-1.
#' @return A list of results, containing: \itemize{
#'         \item{"matches", the matrix of matched treated and control patients, their mean predilection score and treatment effect estimate}
#'         \item{"predilection.score.model", the fitted glmnet object that models the predilection score}
#'         \item{"predilection.scores", the vector of prevalidated predilection scores for each patient}
#'         \item{"model", the sweet spot model, a list containing:\itemize{
#'             \item{the start and end index of the sweet spot}
#'             \item{the bootstrapped p-value}
#'             \item{the mean inside and outside the sweet spot, before and after debiasing}
#'             \item{the distribution around the start and end indices, from the debiasing bootstrap}
#'             \item{the maximum statistic used to compute the p-value}
#'             }}
#' }
#' 
#' @include match_individuals.R
#' @include predilection_scores.R
#' @include find_sweetspot.R
#' @include plot_result.R
#' @include aquamat_data.R
#' 
#' @importFrom glmnet glmnet 
#' @importFrom optmatch pairmatch
#' @import doParallel
#' @importFrom foreach %dopar% foreach
#' @importFrom stats smooth.spline predict
#' @importFrom parallel detectCores
#' 
#' @examples 
#' # Example data with a sweet spot
#' # We generate data with a sweetspot in the middle 20% of predilection scores.
#' # Outside the sweet spot, the treatment effect is 5%.
#' # Inside the sweet spot, the treatment effect is 25%.
#' set.seed(1234)
#' n <- 500; p <- 10;
#' treated    <- sample(c(0,1), n, replace=TRUE)
#' covariates <- matrix(rnorm(n * p), nrow=n, ncol=p)
#' beta       <- rnorm(p)
#' outcome.prob <- 1/(1+exp(-(covariates %*% beta)))
#' in.sweet.spot <- !is.na(cut(outcome.prob, c(.4, .6)))
#' outcome.prob[treated==1] <- outcome.prob[treated==1] + .05
#' outcome.prob[treated==1 & in.sweet.spot] <- outcome.prob[treated==1 & in.sweet.spot] + .2
#' outcome.prob     <- pmin(outcome.prob, 1)
#' positive.outcome <- rbinom(n, 1, prob=outcome.prob)
#' negative.outcome <- 1-positive.outcome
#' 
#' # We limit the number of trials run only to illustrate how to use this function.
#' # In practice, we recommend more bootstrap trials.
#' result <- sweetspot(treated, covariates, 
#'                     negative.outcome, positive.outcome, 
#'                     "binomial", ntrials.bias=250, ntrials.significance=250)
#' plot_sweetspot(result, title="Sweet spot on simulated data")
#' 
#' 
#' # Example data without a sweet spot. 
#' # The overall treatment effect is 5%.
#' set.seed(1234)
#' n <- 500; p <- 10;
#' treated    <- sample(c(0,1), n, replace=TRUE)
#' covariates <- matrix(rnorm(n * p), nrow=n, ncol=p)
#' beta       <- rnorm(p)
#' outcome.probability <- 1/(1+exp(-(covariates %*% beta)))
#' outcome.probability[treated == 1] <- pmin(outcome.probability[treated == 1] + .05, 1) 
#' positive.outcome <- rbinom(n, 1, prob=outcome.probability)
#' negative.outcome <- 1-positive.outcome
#' 
#' result <- sweetspot(treated, covariates, 
#'                     negative.outcome, positive.outcome, 
#'                     "binomial", ntrials.bias=250, ntrials.significance=250)
#' plot_sweetspot(result, title="Sweet spot on simulated data")
#' 
#' @export sweetspot

sweetspot <- function(treated, covariates, negative.outcome, positive.outcome, family, 
                      regularized = F, control.treated.ratio = 1, predilection.score.nfolds = 10,
                      ntrials.significance=1000, ntrials.bias=1000, ncores=NULL){
  
  # check types
  if(!((class(covariates)[1] == "matrix") & (typeof(covariates) == "double"))) { 
    message("The covariates are not a numeric matrix."); stop()
  } 
  
  if(!(family %in% c("binomial", "gaussian"))){ message("family should be 'binomial' or 'gaussian'."); stop()}
   
  if(!is.null(ncores)){
    if(!is.numeric(ncores)) { message("Number of cores should be an integer or NULL."); stop() }
    if(floor(ncores) != ncores)  { message("Number of cores should be an integer or NULL."); stop() }
  }
  if(!is.numeric(c(ntrials.significance, ntrials.bias))){ message("Number of trials should be integers."); stop() }
  if(floor(ntrials.significance) != ntrials.significance) { message("ntrials.significance should be an integer."); stop() }
  if(floor(ntrials.bias) != ntrials.bias) { message("ntrials.bias should be an integer."); stop() }
  
  # compute the predilection scores using prevalidation:
  predilection.scores <- predilection_scores(treated, 
                                             covariates, 
                                             negative.outcome,
                                             family = family, 
                                             regularized = regularized, 
                                             nfolds  = predilection.score.nfolds)
  predilection.score.model = predilection.scores$model
  predilection.scores      = predilection.scores$scores
  
  # match individuals using predilection.scores:
  matches <- match.individuals(treated, predilection.scores, control.treated.ratio, 
                                           positive.outcome, function(treated, control) treated - mean(control))
  
  # find the sweet spot, estimate the p-value, and debias the CATE.
  model <- find.sweetspot(matches[, "apparent_benefit"], 
                          ntrials.significance = ntrials.significance, 
                          ntrials.bias = ntrials.bias,
                          ncores = ncores)

  return(list(
              matches = matches,
              predilection.score.model = predilection.score.model,
              predilection.scores = predilection.scores,
              model = model
  ))
}


R/plot_result.R
#' Visualize the sweet spot analysis on clinical trial data. 
#' 
#' @name plot_sweetspot
#' @param sweetspot_result The result from a sweetspot analysis.
#' @param title The title for your plot.
#' @return A ggplot.
#' 
#' @import ggplot2
#' 
#' @export plot_sweetspot

library(ggplot2)

GREY   <- "#E5E5E5"
BLUE   <- "#62CBD7" 
GREEN1 <- "#9ED393"
GREEN2 <- "#55763C"
SPLINEGREY <- "#333333"
BASE_SIZE  <- 15

plot_sweetspot <- function(sweetspot_result, title=""){
  model    <- sweetspot_result$model
  matches  <- sweetspot_result$matches
  
  smoothed <- smooth.spline(matches[, "mean_score"], matches[, "apparent_benefit"])
    
  n     <- nrow(matches)
  
  zero  <- matches[1, "mean_score"]
  lrge  <- matches[n, "mean_score"]
  start <- matches[model$start.index, "mean_score"]
  end   <- matches[model$end.index,   "mean_score"]
  height.base <- abs(0.1*(max(model$mean.outside, model$mean.outside.debiased) - min(model$mean.inside, model$mean.inside.debiased)))
  base <- min(0, model$mean.outside-height.base, model$mean.outside.debiased-height.base)
  
  p <- ggplot() 
  p <- p + geom_ribbon(aes(x=c(zero, zero, start, start, end, end, lrge, lrge), 
                           ymin=rep(base, 8),
                           ymax=c(base, model$mean.outside, model$mean.outside, model$mean.inside, model$mean.inside, model$mean.outside, model$mean.outside, base),
                           fill="original"), alpha=.8)
  p <- p + geom_ribbon(aes(x=c(zero, zero, start, start, end, end, lrge, lrge), 
                           ymin=rep(base, 8),
                           ymax=c(base, model$mean.outside.debiased, model$mean.outside.debiased, model$mean.inside.debiased, model$mean.inside.debiased, model$mean.outside.debiased, model$mean.outside.debiased, base),
                          fill="debiased"), alpha=.4)
  p <- p + geom_line(aes(x=smoothed$x, y=smoothed$y, color="smoothed\ntreatment effect"), size=.6)
  p <- p + xlim(zero, lrge)
  p <- p + scale_color_manual(breaks=c("debiased", "original", "start", "end", "smoothed\ntreatment effect"), values=c(BLUE, GREY, GREEN1, GREEN2, SPLINEGREY))
  p <- p + scale_fill_manual( breaks=c("debiased", "original", "start", "end"), values=c(BLUE, GREY, GREEN1, GREEN2))
  p <- p + geom_rug(aes(x=matches[sample(model$start.indices, min(n, 1000)), "mean_score"], color="start"), size=1, length = unit(0.05, "npc"), alpha=.1)
  p <- p + geom_rug(aes(x=matches[sample(model$end.indices, min(n, 1000)), "mean_score"],   color="end"),   size=1, length = unit(0.05, "npc"), alpha=.1)
  p <- p + labs(x = "Predilection score", y = "Treatment effect estimate", x="", y="", color="", fill="", 
                subtitle=paste0("p-value: ", model$p.value), title=title) 
  p <- p + theme_bw(base_size=BASE_SIZE) 
  p <- p + theme(
              axis.line = element_line(colour = "#d5d5d5"),
              panel.grid.major.y = element_line(linetype = "dotted", color="#d5d5d5"),
              panel.grid.major.x = element_line(linetype = "dotted", color="#d5d5d5"),
              panel.grid.minor = element_blank(),
              panel.border = element_blank(),
              panel.background = element_blank()) 
  p
}


R/find_sweetspot.R
# library(doParallel)
# library(foreach)
# library(parallel)

# This function computes all differences in means, for a fixed window size k.
# It returns the value of the largest statistic, and the window to which it corresponds
one.statistic <- function(k, benefits.cumsum, n, overall.mean, maxonly=F){
  result  <- (benefits.cumsum[((2+k):n)-1] - benefits.cumsum[(2:(n-k))-1]) - k*overall.mean
  best.ix <- which.max(result)
  return(
    c(1+best.ix, k+best.ix, max(result), k)
  )
}


compute.all.statistics <- function(n, benefit.cs, start, end, maxonly=F){
  result <- do.call(rbind,  lapply(start:(end-1), function(k) one.statistic(k, benefit.cs, n, benefit.cs[n]/n)))
  if(maxonly) return( max(result[, 3]) )

  best   <- result[which.max(result[, 3]), ]

  return(
    list(start.index  = best[1],
         end.index    = best[2],
         mean.inside  = (benefit.cs[best[2]] - benefit.cs[best[1]-1])/best[4],
         mean.outside = (benefit.cs[n] - benefit.cs[best[2]] + benefit.cs[best[1]-1])/(n - best[4]),
         statistic    = best[3]
    )
  )
}

find.sweetspot <- function(observed.benefit, ntrials.significance=1000, ntrials.bias=1000,
                           min.size.fraction = 1/20, max.size.fraction = 1, ncores=NULL){
  
  # https://stackoverflow.com/questions/50571325/r-cran-check-fail-when-using-parallel-functions
  chk <- Sys.getenv("_R_CHECK_LIMIT_CORES_", "")
  if (nzchar(chk) && chk == "TRUE") {
    # use 2 cores in CRAN/Travis/AppVeyor
    cores <- 2L
  } else {
    cores <- if(is.null(ncores)){ detectCores()-1 } else { ncores }
  }
  registerDoParallel(cores = cores)
  
  n                <- length(observed.benefit)
  search.start     <- max(4, floor(min.size.fraction*n))
  search.end       <- floor(max.size.fraction*n)

  # Compute the statistic for all possible windows on our observed data, and choose the largest:
  best.window      <- compute.all.statistics(n, cumsum(observed.benefit), search.start, search.end)

  # Compute the p-value
  pvalue <- foreach(i=1:ntrials.significance, .combine = sum, .inorder=F, .multicombine=T, .maxcombine=ntrials.significance) %dopar% {
    compute.all.statistics(n, cumsum(sample(observed.benefit, replace=F)), search.start, search.end, maxonly=T) >= best.window$statistic
  }

  # Debias:
  values.inside  <- observed.benefit[best.window$start.index:best.window$end.index]
  values.outside <- observed.benefit[-(best.window$start.index:best.window$end.index)]

  debias.boot <- foreach(i=1:ntrials.bias, .combine = rbind, .inorder=F, .multicombine=T, .maxcombine=ntrials.bias) %dopar% {
    compute.all.statistics(n,
                           cumsum(c(sample(values.outside, best.window$start.index - 1, replace=T),
                                    sample(values.inside,  best.window$end.index - best.window$start.index + 1, replace=T),
                                    sample(values.outside, max(0, n - best.window$end.index), replace=T))),
                           search.start, search.end, maxonly=F)
  }


  # Return:
  list(start.index  = best.window$start.index,
       end.index    = best.window$end.index,
       statistic    = best.window$statistic,
       mean.inside  = best.window$mean.inside,
       mean.outside = best.window$mean.outside,

       mean.inside.debiased  = 2 * best.window$mean.inside  - 1/ntrials.bias * sum(unlist(debias.boot[, "mean.inside"])),
       mean.outside.debiased = 2 * best.window$mean.outside - 1/ntrials.bias * sum(unlist(debias.boot[, "mean.outside"])),

       # means         = unname(unlist(boot[, "mean.inside"])),
       start.indices = unname(unlist(debias.boot[, "start.index"])),
       end.indices   = unname(unlist(debias.boot[, "end.index"])),

       p.value      = pvalue/ntrials.significance)
}


R/match_individuals.R
# library(optmatch)

###################################################################################################
# Matching individuals
#
# We use optmatch::pairmatch to perform optimal matching to pair controls with treated individuals.
# We take the following arguments:
# 1. treated: a binary vector with one entry for every individual in the trial. The value 1 indicates that the individual was treated, 
#  and the value 0 indicates that they were a control.
# 2. predilection.scores: a numeric vector containing predilection scores.
# 3. control.treated.ratio: the integer number of controls for each treated individual.
# 4. response: the response for each individual.
# 5. apparent.benefit.function The function to use to compute apparent benefit for each group.
###################################################################################################
match.individuals <- function(treated, predilection.scores, control.treated.ratio, response, apparent.benefit.function){
  df <- data.frame(trt=treated, score=predilection.scores, name=1:length(treated))
  df[df$trt == 1, "name"] = paste0("treated_", df[df$trt == 1, "name"])
  df[df$trt == 0, "name"] = paste0("control_", df[df$trt == 0, "name"])

  # Match treated individuals with controls, and remove those we could not match
  pm <- pairmatch(trt ~ score, data=df, controls=control.treated.ratio, remove.unmatchables=T)
  pm <- pm[!is.na(pm)]

  # Compute information for each match (mean predilection score, apparent benefit)
  get_group_info <- function(group){
    gp <- df$name[as.numeric(names(pm[pm == group]))]

    # Index of controls and treated:
    ctrls <- as.numeric(gsub("control_", "", gp[grepl("control_", gp)]))
    trtds <- as.numeric(gsub("treated_", "", gp[grepl("treated_", gp)]))
    
    c(ctrls, trtds,                                                          # id of each control/treated
      predilection.scores[ctrls],                                            # predilection score
      predilection.scores[trtds],
      mean(c(predilection.scores[ctrls], predilection.scores[trtds])),                # mean score for the group
      apparent.benefit.function(treated = response[trtds], control = response[ctrls]) # apparent benefit for the group
    )
  }
  all.matches <- t(sapply(unique(pm), get_group_info))

  ctrl.names <- if(control.treated.ratio == 1) {"control"} else {paste0("control_", 1:control.treated.ratio)}
  colnames(all.matches) <- c(ctrl.names, "treated", paste0(ctrl.names, "_score"), "treated_score", "mean_score", "apparent_benefit")

  # return the matches in order of increasing predilection score
  all.matches[order(all.matches[, "mean_score"]), ]
}

R/predilection_scores.R
library(glmnet)

####################################################################################################################
# Compute predilection scores.
#
# Compute predilection scores on clinical trial data, using pre-validation. 
# To compute the predilection score, we use Cox regression (survival response), linera regerssion (continuous response),
# or logistic regression (binary response).
# 1. To compute predilection scores for the controls, we use pre-validation: we partition the data into `nfolds` folds, we train a model on `nfolds-1` folds, and compute the score on the remaining fold. 
#    We do this to avoid introducing bias: each individual's score is derived from other individuals in the data.
# 2. To compute predilections cores for the the treated individuals, we use a model trained a model on all controls.
# 
# 
# param treated A binary vector with one entry for every individual in the trial. The value 1 indicates that the individual was treated, 
#  and the value 0 indicates that they were a control.
# param covariates A numeric matrix containing a row of covariates for every individual in the trial (in the same order as in `treated`).
# param response The response we use to compute our predilection score. This will determine the model we use.
# This may be a binary (logistic regression) vector, a continuous (linear regression) vector, or a `Surv` object (Cox proportional hazards).
# param family The response type: one of "Cox", "binomial" or "gaussian".
# param regularized Logical flag for whether to use regularization when computing scores. Default is `FALSE`. If `TRUE`, we predict using 
# param nfolds The (positive integer) number of folds we use in pre-validation. The default is 10 folds.
# returns A vector of predilection scores, one for every individual in the data.
####################################################################################################################

predilection_scores <- function(treated, covariates, response, family = "binomial", regularized = F, nfolds = NULL){
  # Argument checking:
  # if((class(response) == "Surv") & (family != "cox")) { message("The response type (Surv) does not match the family. Use `family = \"cox\" for survival outcomes."); stop()} 
  if((class(response) == "Surv") & (family != "cox")) { message("The response type (Surv) is not binary or continuous."); stop()} 
  if((family == "binomial") & (length(unique(response)) > 2)) { message("The response type has more than two unique values. Use `family = \"gaussian\" for continuous outcomes."); stop()} 
  if(!is.null(nfolds)){
    if(nfolds > sum(treated == 0)) { message("Too many folds for prevalidating predilection scores. Make sure the number of folds is less than or equal to the number of controls."); stop() } 
    if(nfolds < 0) { message("Number of folds for prevalidation should be greater than zero."); stop() }
  }
  
  nf <- if(is.null(nfolds)) {sum(treated == 0)} else {nfolds}
  
  # Fit and predict:
  fit.model <- glmnet::cv.glmnet(covariates[treated == 0, ], 
                         response[treated == 0], 
                         family=family, 
                         nfolds=nf,
                         grouped=F,
                         # intercept=T,
                         keep=T)
  
  s         <- if(regularized) { fit.model$lambda.1se } else {  min(fit.model$lambda)  }  
  
  scores               <- rep(NA, nrow(covariates))
  scores[treated == 0] <- fit.model$fit.preval[, which(fit.model$lambda == s)]
  scores[treated == 1] <- predict(fit.model, s=s, newx=covariates[treated == 1, ])
  
  return(list(model = fit.model, scores = scores))
}


R/aquamat_data.R
#' @title Example randomized controlled trial data
#'
#' Data from a randomized controlled trial comparing artesunate to quinine in the treatment of severe falciparum malaria in African children, with outcome in-hospital mortality.
#' Missing values have been imputed with the median.
#'
#' @docType data
#'
#' @usage data(aquamat)
#'
#' @format A dataframe with the AQUAMAT data.
#'
#' @keywords datasets
#'
#' @references 
#' Dondorp, Arjen M., et al. "Artesunate versus quinine in the treatment of severe falciparum malaria in African children (AQUAMAT): an open-label, randomised trial." The Lancet 376.9753 (2010): 1647-1657.
#' (\href{https://www.sciencedirect.com/science/article/pii/S0140673610619241}{The Lancet})
#' Watson, Leopold et al. Collider bias and the apparent protective effect of glucose-6-phosphate dehydrogenase deficiency on cerebral malaria eLife (2019).
#' (\href{https://elifesciences.org/articles/43154}{eLife})
#' Leopold, Watson et al. Investigating causal pathways in severe falciparum malaria: a pooled retrospective analysis of clinical studies In Press, PLoS Medicine (2019).
#' (\href{https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1002858}{PLoS Medicine})
#' Watson, J.A., Holmes, C.C. Graphing and reporting heterogeneous treatment effects through reference classes. Trials 21, 386 (2020).
#' (\href{ https://doi.org/10.1186/s13063-020-04306-1}{Trials})
#' Craig, E., Redelmeier, D.A., Tibshirani, R. J. Finding and assessing treatment effect sweet spots in clinical trial data. Arxiv (2020).
#' (\href{https://arxiv.org/abs/2011.10157}{arXiv})
#'
#' @source \href{https://github.com/Stije/SevereMalariaAnalysis}{Github}
#'
#' @examples
#' data(aquamat)
#' covariates <- as.matrix(aquamat[, 1:(ncol(aquamat)-3)])
#' treated    <- aquamat$treated
#' negative.outcome <- aquamat$negative.outcome
#' positive.outcome <- aquamat$positive.outcome
#' \donttest{result <- sweetspot(treated, covariates, negative.outcome, positive.outcome, "binomial")}
#' \donttest{plot_sweetspot(result, title="Sweet spot analysis: AQUAMAT trial")}
"aquamat"